### DISCUSSION TITLE
STL for meep

### DISCUSSION THREAD

**Discussion by warsea94** [NanoComp/meep#2627] [Q&A]
Hi to the meep community,

We are currently working on simulating a complex geometry of a plasma chamber. Going through the documentation, we found that the geometries are built up using primitives such as sphere, cylinder etc. 
We would love to know if there is a way or a function which helps in importing the geometry from an STL (Standard Triangle Language) file to get the simulation geometry.

Thanks in advance.

**Comment by smartalecH** [comment]
There's no "built-in" support for structures like this. But meep has a flexible API that lets you do whatever you want, really. 

You can supply an [`epsilon_func`](https://meep.readthedocs.io/en/latest/Python_User_Interface/#geometricobject) as a material for an arbitrary object (eg a block encompassing a region of your simulation domain). This function will expect a `Vector3` point as an input, and a dielectric constant as an output. Meep will effectively loop through all the relevant points and call this function at each point. 

The trick, then, is to cleverly check if that point is inside your stl object within that function. There are many python libraries you can use to pull the vertices from an stl, and then there are various "point in polyhedron" algorithms you can use to formulate the logic.
